File: app/src/main/java/msr/mirudl/app/DownloadEntryStore.java
package msr.mirudl.app;

import msr.mirudl.shared.model.DownloadRecord;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public final class DownloadEntryStore {
    private static final String PREF = "mirudl_downloads";
    private static final String KEY = "entries";

    public enum DeleteScope {
        RECORD_ONLY,
        FILE_ONLY,
        RECORD_AND_FILE
    }

    private DownloadEntryStore() {}

    public static List<DownloadRecord> all(Context context) {
        List<DownloadRecord> items = new ArrayList<>();
        String raw = context.getSharedPreferences(PREF, Context.MODE_PRIVATE)
                .getString(KEY, "[]");
        try {
            JSONArray arr = new JSONArray(raw);
            for (int i = 0; i < arr.length(); i++) {
                JSONObject json = arr.getJSONObject(i);
                DownloadRecord e = new DownloadRecord();
                e.uri = json.optString("uri", "");
                e.filePath = json.optString("filePath", "");
                e.title = json.optString("title", "");
                e.parent = json.optString("parent", "");
                e.size = json.optLong("size", 0);
                e.completedAt = json.optLong("completedAt", 0);
                items.add(e);
            }
        } catch (Exception ignored) {}
        items.sort((a, b) -> Long.compare(b.completedAt, a.completedAt));
        return items;
    }

    public static void add(Context context, DownloadRecord entry) {
        if (entry == null) return;
        List<DownloadRecord> items = all(context);
        String key = entry.key();
        boolean replaced = false;
        for (int i = 0; i < items.size(); i++) {
            if (key.equals(items.get(i).key())) {
                items.set(i, entry);
                replaced = true;
                break;
            }
        }
        if (!replaced) items.add(0, entry);
        save(context, items);
    }

    public static void remove(Context context, DownloadRecord entry) {
        if (entry == null) return;
        String key = entry.key();
        List<DownloadRecord> out = new ArrayList<>();
        for (DownloadRecord e : all(context)) {
            if (!key.equals(e.key())) out.add(e);
        }
        save(context, out);
    }

    public static void removeAll(Context context, List<DownloadRecord> entries) {
        if (entries == null || entries.isEmpty()) return;
        Set<String> keys = new HashSet<>();
        for (DownloadRecord e : entries) keys.add(e.key());
        List<DownloadRecord> out = new ArrayList<>();
        for (DownloadRecord e : all(context))
            if (!keys.contains(e.key())) out.add(e);
        save(context, out);
    }

    private static void save(Context context, List<DownloadRecord> items) {
        JSONArray arr = new JSONArray();
        for (DownloadRecord e : items) {
            JSONObject json = new JSONObject();
            try {
                json.put("uri", e.uri != null ? e.uri : "");
                json.put("filePath", e.filePath != null ? e.filePath : "");
                json.put("title", e.title != null ? e.title : "");
                json.put("parent", e.parent != null ? e.parent : "");
                json.put("size", e.size);
                json.put("completedAt", e.completedAt);
                arr.put(json);
            } catch (Exception ignored) {}
        }
        context.getSharedPreferences(PREF, Context.MODE_PRIVATE)
                .edit().putString(KEY, arr.toString()).apply();
    }
}

File: shared/src/commonMain/kotlin/msr/mirudl/shared/storage/DownloadEntryStore.kt
package msr.mirudl.shared.storage

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import msr.mirudl.shared.model.DownloadRecord

object DownloadEntryStore {

    private const val KEY = "entries"

    @Serializable
    private data class JsonRecord(
        val uri: String = "",
        val filePath: String = "",
        val title: String = "",
        val parent: String = "",
        val size: Long = 0L,
        val completedAt: Long = 0L
    )

    private val json = Json { ignoreUnknownKeys = true }

    fun all(storage: AppStorage): List<DownloadRecord> {
        val raw = storage.getString(KEY, "[]")
        return try {
            val arr = json.decodeFromString<List<JsonRecord>>(raw)
            arr.map { j ->
                DownloadRecord(
                    uri = j.uri.ifEmpty { null },
                    filePath = j.filePath.ifEmpty { null },
                    title = j.title.ifEmpty { null },
                    parent = j.parent.ifEmpty { null },
                    size = j.size,
                    completedAt = j.completedAt
                )
            }.sortedByDescending { it.completedAt }
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun add(storage: AppStorage, entry: DownloadRecord?) {
        if (entry == null) return
        val items = all(storage).toMutableList()
        val key = entry.key()
        val idx = items.indexOfFirst { it.key() == key }
        if (idx >= 0) items[idx] = entry else items.add(0, entry)
        save(storage, items)
    }

    fun remove(storage: AppStorage, entry: DownloadRecord?) {
        if (entry == null) return
        val key = entry.key()
        val out = all(storage).filter { it.key() != key }
        save(storage, out)
    }

    fun removeAll(storage: AppStorage, entries: List<DownloadRecord>?) {
        if (entries.isNullOrEmpty()) return
        val keys = entries.mapNotNull { it.key() }.toSet()
        val out = all(storage).filter { it.key() !in keys }
        save(storage, out)
    }

    private fun save(storage: AppStorage, items: List<DownloadRecord>) {
        val arr = items.map { e ->
            JsonRecord(
                uri = e.uri ?: "",
                filePath = e.filePath ?: "",
                title = e.title ?: "",
                parent = e.parent ?: "",
                size = e.size,
                completedAt = e.completedAt
            )
        }
        storage.setString(KEY, json.encodeToString(arr))
    }
}
